test_runner/regress/test_tenant_detach.py

```diff
@@ -108,6 +108,218 @@ def test_tenant_reattach(
+        current_lsn = Lsn(query_scalar(cur, "SELECT pg_current_wal_flush_lsn()"))
+
+    main_endpoint.stop()
+    env.neon_cli.create_branch("child", tenant_id=tenant_id, ancestor_start_lsn=current_lsn)
+
+    pageserver_http.tenant_detach(tenant_id)
+
+    # This fail point will pause the WAL ingestion on the main branch, after the
+    # the first insert
+    pageserver_http.configure_failpoints([("wal-ingest-logical-message-sleep", "return(5000)")])
+
+    # Attach and wait a few seconds to give it time to load the tenants, attach to the
+    # safekeepers, and to stream and ingest the WAL up to the pause-point.
+    before_attach_time = time.time()
+    pageserver_http.tenant_attach(tenant_id)
+    time.sleep(3)
+
+    # The wal ingestion on the main timeline should now be paused at the fail point.
+    # Run a query on the child branch. The GetPage requests for this should recurse to the
+    # parent timeline, and wait for the WAL to be ingested there. Otherwise it won't see
+    # the second insert.
+    child_endpoint = env.endpoints.create_start("child", tenant_id=tenant_id)
+    with child_endpoint.cursor() as cur:
+        cur.execute("SELECT * FROM test_reattach")
+        assert cur.fetchall() == [("before pause",), ("after pause",)]
+
+    # Sanity check that the failpoint was reached
+    assert env.pageserver.log_contains('failpoint "wal-ingest-logical-message-sleep": sleep done')
+    assert time.time() - before_attach_time > 5
+
+    # Clean up
+    pageserver_http.configure_failpoints(("wal-ingest-logical-message-sleep", "off"))
+
+
+# Detach and re-attach a tenant with a branch
+#
+# This exercises the same race condition after tenant attach as
+# 'test_tenant_with_branches_reattach_slow_replay', but this test case
+# is closer to the original scenario where we originally found the
+# issue.
+#
+# In this scenario, the incorrect result to get-request leads to
+# *permanent damage* in the child timeline, because ingesting the WAL
+# on the child timeline depended on incorrect view of the parent. This
+# test reproduced one such case; the symptom was an error on the child, when
+# trying to connect to the child endpoint after re-attaching the tenant:
+#
+# FATAL: database "neondb" does not exist
+#
+# In the original case where we bumped into this, the error was slightly
+# different:
+#
+# FATAL:  "base/16385" is not a valid data directory
+# DETAIL:  File "base/16385/PG_VERSION" is missing.
+#
+### Detailed explanation of the original bug and why it lead to that error:
+#
+# The WAL on the main and the child branches look like this:
+#
+#    Main                                  Child
+# 1. CREATE DATABASE
+#           <child branch is created>
+# 2. CREATE TABLE AS SELECT ...            3. CREATE TABLE AS SELECT ...
+#
+# None of these WAL records have been flushed to disk or uploaded to remote
+# storage in the pageserver yet, when the tenant is detached.
```

Comment:
I think any freeze, flush, and wait uploads to complete, will be fine, because we are hoping that they are not yet on pageservers in-memory layers. I guess I'd prefer to have more asserts that `remote_consistent_lsn` is `initdb_lsn` (probably cannot be 0).

Reply:
> IMO regular detach should freeze & flush, no?

/me reads the code

Huh, it does not. Detach calls `task_mgr::shutdown_tasks`, which requests the flush_loop to exit. Pageserver shutdown does freeze and flush, but not detach.

Reply:
I think changing that is out of scope for this PR. But I'll change the test so that instead of detaching the tenant gracefully, it kills the pageserver and removes the tenant directory. That simulates attaching the tenant to a different pageserver, after a hard, unrecoverable crash of the original pageserver.

Reply:
Opened https://github.com/neondatabase/neon/issues/4219 to track changing tenant_detach behavior

Reply:
> I think any freeze, flush, and wait uploads to complete, will be fine, because we are hoping that they are not yet on pageservers in-memory layers. I guess I'd prefer to have more asserts that `remote_consistent_lsn` is `initdb_lsn` (probably cannot be 0).

To avoid the original bug, the WAL needs to be uploaded to remote storage. Otherwise, after re-attach the `remote_consistent_lsn` won't include it, and `get_page_reconstruct_data` won't see it, unless it waits for it to arrive.

Can you elaborate what assert you'd like to see, and where?

Reply:
> As in if we had uploaded the layers to remote storage, then we wouldn't be walreceiving them anymore, we would just download the layers on-demand which is expected to work.

Correct!

Reply:
Hmm, I'm not sure what value I would expect `remote_consistent_lsn` to have in the child here. The 'index_part.json` hasn't been uploaded at all yet, so even the existence of the timeline isn't durable in remote storage yet. But if you ignored that, then I guess the current behaviour is correct. Another valid answer IMO would be `remote_consistent_lsn==ancestor_start_lsn`.

https://github.com/neondatabase/neon/pull/4204 makes this more clear. With that, I think the correct answer here is `remote_consistent_lsn==ancestor_start_lsn`. Maybe add that assertion as part of that PR.

Reply:
> `remote_consistent_lsn==ancestor_start_lsn`

Aah this is interesting, I didn't see it but obviously it should be like this. The asserts worked, but now I don't understand why. The timeline is branched off the lsn which should had been awaited for. 

I already merged the PR so I am doing a new one to clear up this confusion (of mine).